### PR TITLE
Implement Shapley values

### PR DESCRIPTION
**Pull request by RAMitchell**
Provides both global and local feature importance metrics.

### CODE REVIEW

**Review by seberg — APPROVED** [review]
Nice, this LGTM (I'll trust you on some math details, but the code calculations are straight forward and look right/reasonable).
Tests seem fine, although I wonder if they could do more check (looks mostly like sanity checks currently).

Basically, all comments are just rambles/nits so "take it if you think it'll be nicer".

The only thing that I would really care about is the `if n_outputs == 1:` logic, but I suspect that a prediction result cannot be 1-D with a single element.

File: legateboost/legateboost.py [hunk 1]
```diff
@@ -444,6 +446,113 @@ def dump_models(self) -> str:
...
+    def global_attributions(
+        self,
+        X: cn.array,
+        y: cn.array,
+        metric: Optional[BaseMetric] = None,
+        random_state: Optional[np.random.RandomState] = None,
+        n_samples: int = 5,
+        assert_efficiency: bool = False,
+    ) -> Tuple[cn.array, cn.array]:
+        """Compute global feature attributions for the model. Global
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply legateboost/legateboost.py:458]: That is really helpful.

File: legateboost/shapley.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            X_temp[:, feature] = X[:, feature]
+            loss = metric.metric(y, predict_fn(X_temp), w)
+            v[feature, i] = loss - previous_loss
+            previous_loss = loss
+
+    for i in range(n_samples):
+        p_a = random_state_.permutation(X.shape[1])
+        p_b = cn.flip(p_a)
+        eval_sample(p_a, v_a)
+        eval_sample(p_b, v_b)
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply legateboost/shapley.py:56]: Yes, every permutation sample uses a different background sample.

File: legateboost/shapley.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    se = cn.std(v, axis=-1, ddof=1) / cn.sqrt(n_samples)
+
+    if assert_efficiency:
+        pred = predict_fn(X)
+        assert cn.allclose(cn.sum(shapley_values, axis=1), pred), (
+            cn.sum(shapley_values, axis=1)[0],
+            pred[0],
+        )
+
+    if n_outputs == 1:
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply legateboost/shapley.py:130]: I think the predict function here guarantees a 2d array.

File: legateboost/shapley.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+    # antithetic sampling
+    # perhaps we can do a running mean/se to avoid the last dimension
+    v_a = cn.zeros((X.shape[0], X.shape[1] + 1, n_outputs, n_samples))
+    v_b = cn.zeros((X.shape[0], X.shape[1] + 1, n_outputs, n_samples))
+
+    def eval_sample(p: cn.array, v: cn.array) -> None:
+        X_temp = X_background[
+            gen.integers(0, X_background.shape[0], n_background_samples * X.shape[0])
+        ].reshape((X.shape[0], n_background_samples, X.shape[1]))
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply legateboost/shapley.py:97]: Didn't know you could do this!

File: legateboost/shapley.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+) -> Tuple[cn.array, cn.array]:
+    def predict_fn(X: cn.array) -> cn.array:
+        fn = model.predict if is_regressor(model) else model.predict_proba
+        p = fn(X)
+        if p.ndim == 1:
+            return p.reshape(-1, 1)
+        return p
+
+    random_state_ = check_random_state(random_state)
+    n_background_samples = 5
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply legateboost/shapley.py:84]: I think its reasonable, as increasing n_samples here will also increase the number of background samples taken. I didn't want to give the user knobs to turn that are difficult to explain.

**Pull request merged**